File: Leetcode/152. Maximum Product Subarray/maximum_product_subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        prefix = 0
        suffix = 0
        res = nums[0]
        n = len(nums)
        for i in range(0, len(nums)):
            prefix = (1 if prefix==0 else prefix)*nums[i]
            suffix = (1 if suffix==0 else suffix)*nums[n-1-i]
            res = max(res, prefix, suffix)
            
        return res
    
    def maxProduct1(self, nums: List[int]) -> int:
        max_product = nums[0]
        min_product = nums[0]
        res = nums[0]
        n = len(nums)
        for i in range(1, len(nums)):
            curr = nums[i]
            if(curr<0):
                max_product,min_product = min_product, max_product
            max_product = max(curr, curr*max_product)
            min_product = min(curr, curr*min_product)
            res = max(res, max_product)
            
        return res

item = Solution()
print(item.maxProduct1([2,3,-2,4]))

File: Leetcode/152. Maximum Product Subarray/maximumProductSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
  let maxProduct = nums[0];
  let res = nums[0];
  let minProduct = nums[0];
  for (let i = 1; i < nums.length; i++) {
    if (nums[i] < 0) {
      [maxProduct, minProduct] = [minProduct, maxProduct];
    }
    maxProduct = Math.max(nums[i], nums[i] * maxProduct);
    minProduct = Math.min(nums[i], nums[i] * minProduct);
    res = Math.max(res, maxProduct);
  }
  return res === -0 ? 0 : res;
};

// Kadane Algorithms

var maxProduct1 = function (nums) {
  let prefix = 0;
  let suffix = 0;
  let res = nums[0];
  let n = nums.length;
  
  for (let i = 0; i < n; i++) {
    // If prefix or suffix hit 0 due to previous element, reset to 1
    prefix = (prefix === 0 ? 1 : prefix) * nums[i];
    suffix = (suffix === 0 ? 1 : suffix) * nums[n - 1 - i];
    
    res = Math.max(res, prefix, suffix);
  }
  
  // Hardened check against Signed Zero (-0)
  return res === -0 ? 0 : res;
};
